engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+		if obj.init.Debug {
+			obj.init.Logf("requesting server rebuild")
+		}
+		if err := obj.rebuildServer(ctx); err != nil {
+			return false, errwrap.Wrapf(err, "rebuildServer failed")
+		}
+	}
+	return checkOK, nil
+}
+
+// checkApplyRescueMode enables/disables rescue mode for the given keyset, to
+// ensure that rescue mode is available over SSH when needed.
+// NOTE: changing image type currently requires a manual re-enable.
+// TODO: add checkapply for rescue mode specs when already enabled?
+func (obj *HetznerVMRes) checkApplyRescueMode(ctx context.Context, apply bool) (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("checkApplyRescueMode(apply: %t)", apply)
+	}
+	// Check if rescue mode is already in the intended configuration.
+	checkOK := true
+	if obj.server.RescueEnabled == obj.rescueEnable {
+		return true, nil
+	}
+	checkOK = false
+	if !apply {
+		return false, nil
+	}
+	// Otherwise, apply the necessary changes.
+	if obj.rescueEnable {
+		if err := obj.getRescueKeys(ctx); err != nil {
+			return false, errwrap.Wrapf(err, "failed to build rescue mode keylist")
+		}
+		opts := hcloud.ServerEnableRescueOpts{
+			Type:    hcloud.ServerRescueType(obj.ServerRescueMode),
+			SSHKeys: obj.rescueKeys,
+		}
+		if obj.init.Debug {
+			obj.init.Logf("requesting server rescue mode enable")
+		}
+		_, _, err := obj.client.Server.EnableRescue(ctx, obj.server, opts)
+		if err != nil {
+			return false, errwrap.Wrapf(err, "failed to enable rescue mode")
+		}
+		// EnableRescue returns a root password, but this is ignored in favor
+		// of connecting to the server in rescue mode over SSH.
+	}
+	if !obj.rescueEnable {
+		if obj.init.Debug {
+			obj.init.Logf("requesting server rescue mode disable")
+		}
+		_, _, err := obj.client.Server.DisableRescue(ctx, obj.server)
+		if err != nil {
+			return false, errwrap.Wrapf(err, "failed to disable rescue mode")
+		}
+	}
+	// Wait until the requested changes are confirmed.
+	if err := obj.waitFor(ctx, rescueModeOK); err != nil {
+		return false, errwrap.Wrapf(err, "waitFor(rescueModeOK) exited early")
+	}
+	return checkOK, nil
+}
+
+// getServerUpdate pings the Hetzner API for up-to-date server info.
+func (obj *HetznerVMRes) getServerUpdate(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("getServerUpdate()")
+	}
+	server, _, err := obj.client.Server.GetByName(ctx, obj.Name())
+	if err != nil {
+		return errwrap.Wrapf(err, "failed serverupdate request")
+	}
+	obj.server = server
+	return nil
+}
+
+// Cmp compares two HetznerVMRes structs. Returns nil if the comparison holds
+// true, and otherwise an error is thrown to identify the difference.
+func (obj *HetznerVMRes) Cmp(r engine.Res) error {
+	// check if empty
+	if obj == nil && r == nil {
+		return nil
+	}
+	if (obj == nil) != (r == nil) {
+		return fmt.Errorf("one resource is empty")
+	}
+	// compare types
+	res, ok := r.(*HetznerVMRes)
+	if !ok {
+		return fmt.Errorf("not a %s", obj.Kind())
+	}
+	// compare resource fields
```

Comment:
I didn't include `waitinterval`, `waittimeout` and `allowdeletion`, all other public params are there. I did it this way because these three only affect the interaction with the API and not the actual server instance, and the definition of Cmp in the resource guide seems to allow omissions in that case. 

Now I'm thinking it's safer to minimize exceptions and just include everything to avoid surprises for any future developments. If I understand correctly this affects whether or not two resources can be merged in a graph? I was unsure about that earlier, but if that's the case then it's probably important to include `AllowDeletion`, and the two time params would prevent some unexpected timing issues.  

I'm also not comparing any of the private params, since those are generated (and some can be updated) at or after initialization. Will keep those out of Cmp, but please let me know if you disagree.